### PR TITLE
Added BoundaryMesher

### PR DESCRIPTION
**Pull request by regislebrun**
This class allows to extract the boundary of a given mesh as a new mesh. Depending on an offset scalar parameter, the boundary mesh can be:
  * A pure surface mesh, it means a (d-1) dimensional mesh embedded into a d dimensional space: the vertices are d-dimensional, the simplices have (d+1) indices but the two last indices are equal, meaning that the simplex is flat and has to be interpreted as a (d-1) simplex corresponding to the dth first indices
  * An actual d dimensional mesh (a *thick* boundary), where each simplex has a face on the boundary and another vertex at the perpendicular of the face center. This new vertex is at distance offset from the face, and is inside of the initial mesh if offset is negative, outside of the initial mesh if offset is positive.

Here is a 2D illustration:
```
import openturns as ot
from openturns.viewer import View
import openturns.experimental as otexp
import matplotlib.pyplot as plt

# A nice level set
function = ot.SymbolicFunction(["x", "y"], ["(x^3+y^2-3)^2-x^2"])
levelSet = ot.LevelSet(function, ot.Less(), 1.5)
# Create the Mesh
mesh2D = ot.LevelSetMesher([21]*2).build(levelSet, ot.Interval([-2.0, -3.0], [2.0, 3.0]))
# Build the Mesh boundary
mesh2DBoundary = otexp.BoundaryMesher().build(mesh2D)
# Build a thick outside Mesh boundary
mesh2DBoundaryOutside = otexp.BoundaryMesher().build(mesh2D, 0.1)
# Build a thick inside Mesh boundary
mesh2DBoundaryInside = otexp.BoundaryMesher().build(mesh2D, -0.1)
# Create a Graph
graph = ot.Graph("", "", "", True, "left")
g = mesh2D.draw()
graph.add(g)
# Then, draw it
fig = plt.figure(figsize=(16, 4))
axis = fig.add_subplot(141)
View(graph, figure=fig, axes=[axis], add_legend=True)
axis.set_xlim(auto=True)

# Create a Graph
graph = ot.Graph("", "", "", True, "left")
g = mesh2DBoundary.draw()
graph.add(g)
# Then, draw it
axis = fig.add_subplot(142)
View(graph, figure=fig, axes=[axis], add_legend=True)
axis.set_xlim(auto=True)

# Create a Graph
graph = ot.Graph("", "", "", True, "left")
g = mesh2DBoundaryInside.draw()
graph.add(g)
# Then, draw it
axis = fig.add_subplot(143)
View(graph, figure=fig, axes=[axis], add_legend=True)
axis.set_xlim(auto=True)

# Create a Graph
graph = ot.Graph("A 2D mesh with its boundary, its inside thick boundary and its outside thick boundary", "", "", True, "left")
g = mesh2DBoundaryOutside.draw()
graph.add(g)
# Then, draw it
axis = fig.add_subplot(144)
View(graph, figure=fig, axes=[axis], add_legend=True)
axis.set_xlim(auto=True)
plt.show()
```
resulting into the following figure:

Figure_1

Along the way, the LevelSetMesher has been significantly improved by providing a new way to project vertices on the level set boundary. For simplices with vertices on both side of the boundary, in master the options are:
  * to use a linear interpolation between each vertex outside of the level set and the center of the vertices inside of the level set when the flag project==false
  * to use a NearestPointAlgorithm to compute the actual projection of each vertex outside of the level set on the boundary when the flag project==true

While the last option has appealing mathematical properties, in practice it is both slow and fragile: from time to time the optimization step fails, resulting into badly shaped simplices. A new version of the projection is proposed in this PR, in the spirit of the linear interpolation. This linear interpolation is replaced by the computation of the actual intersection between the ray and the boundary. If it fails, then the usual projection is used. This behaviour is controlled by the project flag and the "LevelSetMesher-SolveEquation" entry in ResourceMap.

One application of the BoundaryMesher class is to be able to sample uniformly the boundary of a domain described by a mesh by sampling a thick boundary using the UniformOverMesh class, with an offset of the order of 1e-8*max(diameter(face))

### CODE REVIEW

File: lib/src/Base/Geom/MeshBoundaryFactory.cxx [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+/* Here is the interface that all derived class must implement */
+
+
+namespace
+{
+struct FaceHasher
+{
+  UnsignedInteger operator()(const Indices &face) const
+  {
```
Comment (jschueller) [inline lib/src/Base/Geom/MeshBoundaryFactory.cxx:66]: maybe add a link to the method used, and a comment why we hash until the n-1 vertex

File: lib/src/Base/Geom/MeshBoundaryFactory.cxx [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  }
+}; // FaceCompare
+
+} // Anonymous namespace
+
+/*
+  Amélioration des performances:
+  1) Etape de préparation des faces.
+  + utiliser une IndicesCollection, pour chaque simplex trier ses numéros de noeuds
+    sans création d'Indices temporaires puis injecter ces numéros de noeuds
```
Comment (jschueller) [inline lib/src/Base/Geom/MeshBoundaryFactory.cxx:96]: english :)

File: python/src/MeshBoundaryFactory_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,56 @@
+%feature("docstring") OT::MeshBoundaryFactory
+"Creation of the boundary mesh of a given mesh..
+
+.. warning::
+    This class is experimental and likely to be modified in future releases.
+    To use it, import the ``openturns.experimental`` submodule.
+
+Available constructor:
+    MeshBoundaryFactory()
```
Comment (jschueller) [inline python/src/MeshBoundaryFactory_doc.i.in:9]: documenting the parameters instead

File: lib/src/Base/Geom/LevelSetMesher.cxx [hunk 4]
```diff
@@ -203,51 +207,74 @@ Mesh LevelSetMesher::build(const LevelSet & levelSet,
...
+                  // Try with the new gradients
+                  try
+                  {
+                    solver_.run();
+                    result = solver_.getResult();
+                  }
+                  catch(...)
+                  {
+                    // There is definitely a problem with this vertex. Try a gradient-free solver
+                    Cobyla solver(solver_.getProblem());
```
Comment (jschueller) [inline lib/src/Base/Geom/LevelSetMesher.cxx:268]: add an attribute to change the solver

**Comment by jschueller** [comment]
What do you think of BoundaryMesher (like LevelSetMesher, IntervalMesher) ? (If you agree there's a rename.sh script)
Currently tests are missing, but otherwise looks good.

File: lib/src/Base/Geom/MeshBoundaryFactory.cxx (hunk 1 above)
Comment (regislebrun) [reply lib/src/Base/Geom/MeshBoundaryFactory.cxx:72]: done

File: lib/src/Base/Geom/MeshBoundaryFactory.cxx (hunk 2 above)
Comment (regislebrun) [reply lib/src/Base/Geom/MeshBoundaryFactory.cxx]: fixed

File: python/src/MeshBoundaryFactory_doc.i.in (hunk 3 above)
Comment (regislebrun) [reply python/src/MeshBoundaryFactory_doc.i.in:9]: And when there is no parameter?

File: lib/src/Base/Geom/LevelSetMesher.cxx (hunk 4 above)
Comment (regislebrun) [reply lib/src/Base/Geom/LevelSetMesher.cxx:268]: It is a fallback solver, no need to change it IMO. BTW, I don't know how to check if the user defined solver can solve an optimization problem with nonlinear constraints.

**Comment by regislebrun** [comment]
@jschueller the initial name I chose was BoundaryMesher, but a mesher is in charge of converting a Domain into a Mesh, here it is an extraction from a Mesh. I also tried BoundaryExtractor but it sounds weird. Hence I went for MeshBoundaryFactory...

**Comment by jschueller** [comment]
I still prefer BoundaryMesher :)

**Comment by regislebrun** [comment]
I was thinking about extending the class to general domains in order to match the new name (BoundaryMesher). The idea is to rely on the Domain::contains() method to build a LevelSet (convert true to 1, false to 0 and solve for a level=0.5). It can be done using a Mesh constructor, but as it involves a significant amount of computation it may be better to create a DomainMesher class. This class could have shortcuts when the domain is a MeshDomain, an Interval or a LevelSet as there are either no work to do or dedicated meshers.
Do you think it is a good idea? Do you have other proposals for the architecture?

**Comment by jschueller** [comment]
no, I dont have other comments, other features probably wait another release though, no hurry,
but I'll be interested in rediscussing meshing stuff for the TruncatedOverMesh (for another release also ...)

**Comment by regislebrun** [comment]
@jschueller What is the issue with TruncatedOverMesh? It looks quite good to me, even if the calibration of the rejection could be optimized (I don't know how right now...). It would be an internal optimization with no change in the API.

**Comment by jschueller** [comment]
the problem was the CDF method not using cubature yet to due to having to compute mesh intersection with the quadrant (still using the old integration method), but yes maybe its good enough to go

**Comment by regislebrun** [comment]
@jschueller Regarding the tests, as I use an std::unordered_map, the order in which the faces of the boundary are extracted depends on the implementation of this container, leading to different orders on Windows and macos. As a consequence, I ignore the output of the test.

File: python/test/CMakeLists.txt [hunk 5]
```diff
@@ -261,6 +261,7 @@ ot_pyinstallcheck_test (LevelSet_std)
 ot_pyinstallcheck_test (LevelSetMesher_std)
 ot_pyinstallcheck_test (Mesh_std)
 ot_pyinstallcheck_test (Mesh_draw IGNOREOUT)
+ot_pyinstallcheck_test (MeshBoundaryFactory_std IGNOREOUT)
```
Comment (jschueller) [inline python/test/CMakeLists.txt:264]: you needs some asserts and remove the .expout

Comment (regislebrun) [reply python/test/CMakeLists.txt]: done

File: python/src/MeshBoundaryFactory_doc.i.in (hunk 3 above)
Comment (regislebrun) [reply python/src/MeshBoundaryFactory_doc.i.in]: done

**Review by jschueller — APPROVED** [review]

**Review by sofianehaddad — APPROVED** [review]

**Pull request merged**